mm2src/mm2_event_stream/src/manager.rs

```diff
@@ -0,0 +1,516 @@
+        self.read().clients.values().for_each(|info| {
+            info.send_event(event.clone());
+        });
+    }
+
+    /// Creates a new client and returns the event receiver for this client.
+    pub fn new_client(&self, client_id: u64) -> Result<mpsc::Receiver<Arc<Event>>, StreamingManagerError> {
+        let mut this = self.write();
+        if this.clients.contains_key(&client_id) {
+            return Err(StreamingManagerError::ClientExists);
+        }
+        // Note that events queued in the channel are `Arc<` shared.
+        // So a 1024 long buffer isn't actually heavy on memory.
+        let (tx, rx) = mpsc::channel(1024);
+        let client_info = ClientInfo::new(tx);
+        this.clients.insert(client_id, client_info);
+        Ok(rx)
+    }
+
+    /// Removes a client from the manager.
+    pub fn remove_client(&self, client_id: u64) -> Result<(), StreamingManagerError> {
+        let mut this = self.write();
+        // Remove the client from our known-clients map.
+        let client_info = this
+            .clients
+            .remove(&client_id)
+            .ok_or(StreamingManagerError::UnknownClient)?;
+        // Remove the client from all the streamers it was listening to.
+        for streamer_id in client_info.listening_to {
+            if let Some(streamer_info) = this.streamers.get_mut(&streamer_id) {
+                streamer_info.remove_client(&client_id);
+            } else {
+                log::error!(
+                    "Client {client_id} was listening to a non-existent streamer {streamer_id}. This is a bug!"
+                );
+            }
+            // If there are no more listening clients, terminate the streamer.
+            if this.streamers.get(&streamer_id).map(|info| info.clients.len()) == Some(0) {
+                this.streamers.remove(&streamer_id);
+            }
+        }
+        Ok(())
+    }
+
+    /// Removes a streamer if it is no longer running.
+    ///
+    /// Aside from us shutting down a streamer when all its clients are disconnected,
+    /// the streamer might die by itself (e.g. the spawner it was spawned with aborted).
+    /// In this case, we need to remove the streamer and de-list it from all clients.
+    fn remove_streamer_if_down(&self, streamer_id: &str) {
+        let mut this = self.write();
+        if let Some(streamer_info) = this.streamers.get_mut(streamer_id) {
+            if streamer_info.is_down() {
+                // Remove the streamer from our registry.
+                if let Some(streamer_info) = this.streamers.remove(streamer_id) {
+                    // And remove the streamer from all clients listening to it.
+                    for client_id in streamer_info.clients {
+                        if let Some(info) = this.clients.get_mut(&client_id) {
+                            info.remove_streamer(streamer_id);
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Too many nested blocks here, which makes things harder to read/understand. Personally, I would prefer following approach instead:

```diff
diff --git a/mm2src/mm2_event_stream/src/manager.rs b/mm2src/mm2_event_stream/src/manager.rs
index 38552836f..6e8f7cb7a 100644
--- a/mm2src/mm2_event_stream/src/manager.rs
+++ b/mm2src/mm2_event_stream/src/manager.rs
@@ -306,17 +306,17 @@ impl StreamingManager {
     /// In this case, we need to remove the streamer and de-list it from all clients.
     fn remove_streamer_if_down(&self, streamer_id: &str) {
         let mut this = self.write();
-        if let Some(streamer_info) = this.streamers.get_mut(streamer_id) {
-            if streamer_info.is_down() {
-                // Remove the streamer from our registry.
-                if let Some(streamer_info) = this.streamers.remove(streamer_id) {
-                    // And remove the streamer from all clients listening to it.
-                    for client_id in streamer_info.clients {
-                        if let Some(info) = this.clients.get_mut(&client_id) {
-                            info.remove_streamer(streamer_id);
-                        }
-                    }
-                }
+        let Some(streamer_info) = this.streamers.get_mut(streamer_id) else { return };
+
+        if !streamer_info.is_down() {
+            return;
+        }
+
+        let Some(streamer_info) = this.streamers.remove(streamer_id) else { return };
+
+        for client_id in streamer_info.clients {
+            if let Some(info) = this.clients.get_mut(&client_id) {
+                info.remove_streamer(streamer_id);
             }
         }
     }
```

Reply:
[3c5b1c6](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2172/commits/3c5b1c6bcec84219ac2d6243649ef584936920c7)